### DISCUSSION TITLE
Is spacy.cli.download.download public API?

### DISCUSSION THREAD

**Discussion by iliakur** [explosion/spaCy#9733] [Help: Best practices]
# Context
At work we wrap spacy and stanza in one codebase that provides uniform setup/access to models from both libraries. Part of this uniform access is downloading the models.
We're currently downloading Stanza models from inside Python and we'd like to do the same for spacy models. We're aware that this isn't exactly recommended by the docs, but we choose this option knowingly due to some other constraints.

# Question
Can we import and invoke `spacy.cli.download.download` to perform the download? Afaict it's essentially a wrapper around `pip install` while taking care of some book-keeping. It also appears pretty stable based on commit history.

**Comment by adrianeboyd** [comment]
The `spacy.cli.download.download` API should be stable. I'm sure we're not quite at 100% on this, but any internal methods that don't start with `_` should be stable even if they're not in the official API docs, and the CLI methods are quite stable in practice.

For this purpose, the main difference between stanza models and spacy models is that spacy models are separate python packages rather than just downloaded directories. Detecting which packages are currently installed in a python environment is expensive, so it's done once when a python script starts and if you install a model package in a running python process, it won't be found by `spacy.load` immediately afterwards. But if your download script is a separate script, this should fine. (Or you'll have to look for workarounds that refresh the internal state about loaded packages, which is possible, but enough of a headache, especially way in the past for python 2, that we haven't tried to support it internally.)

**Comment by iliakur** [comment]
Downloading and installing won't happen in the same process. However, we have a different wrinkle: we are mounting the models into a docker container. Which means we need to a) tell pip to install into a custom directory, b) tell `spacy.load` to load from the same directory. I guess b) can be achieved by tracking the full path to the downloaded model and passing that to `spacy.load`.

**Reply by adrianeboyd** [reply]
I think you can tell `pip` install in a custom directory and then add that to `PYTHONPATH` in the container? It's not something I do a lot so I'm not entirely sure of the details off the top of my head, but something similar should be possible.

Loading from the package name is better than loading from the path in general because if there's custom code, it's loaded automatically by `spacy.load`. There's no custom code in the provided models, but be aware of it for other models.

**Reply by iliakur** [reply]
Indeed there's the `--target` option to `pip install`. Thanks for the point about loading from path vs using `PYTHONPATH`, I'll take this into account!

**Reply by iliakur** [reply]
I just realized this is trivial to do inside a running Python process! So something like this should work, right?

```python
import sys
import spacy

sys.path.append('our/custom/path')  # this will be done in a safer way in prod
spacy.load('en_core_web_sm')
```

**Reply by polm** [reply]
You should just be able to do `spacy.load("out/custom/path/en_core_web_sm")`.

**Reply by adrianeboyd** [reply]
I'm not sure exactly when `importlib.metadata` is going to update its list of packages, so I'm not sure whether that would work. It might work if you don't load anything before modifying the path?

**Reply by adrianeboyd** [reply]
We're trying to load from the package name because that's a better generic practice.

**Reply by polm** [reply]
Ah, my mistake, sorry about that.

**Reply by iliakur** [reply]
So that's a *yes* to my question, right? :)

**Reply by adrianeboyd** [reply]
My answer is: that might work. You'd need to test it carefully, and be aware that it might depend a bit on the python version because of internal details in `importlib.metadata`.

**Reply by iliakur** [reply]
Understood, thanks! I'll report here with my findings once we get it to work.

**Reply by iliakur** [reply]
Hmm, seems like I've stumbled across a bug. `pip_args` are completely ignored. This is because they are declared as variadic positional arguments **after** keyword arguments. Some examples to illustrate what I mean:

```python
download(model, direct=True, sdist=False, "--no-deps") # -> `pip_args` is empty
download(model, "--no-deps", direct=True, sdist=False) # -> Error: multiple values for `direct`
download(model, "--no-deps") # -> `direct="--no-deps"
```

The fix for this was very easy, I just moved `*pip_args` before `direct` and `sdist` in the function declaration. I've noticed, however, that that corner of the code is somewhat under-tested. Getting it under test would require either some refactoring or willingness to perform a model download during testing (possible performance hit).

How do you prefer that we proceed? Should I just open a PR with the basic fix and follow up with some tests later, or should I also do the tests/refactoring?

**Reply by adrianeboyd** [reply]
Hmm, I'm sure this works from the command-line. Ah, I think it will work if you specify `direct` and `sdist` as args instead of kwargs, so:

```python
download(model, True, False, "--no-deps")
```

This is definitely a bit weird, but I'd have to think a bit about the best way to modify this without breaking anything.

We do have some tests in the CI for the CLI commands that we don't have directly the unit tests because of the model downloads, which we restrict to one CI instance (just by python version, it's mildly hacky but works to test the compatibility table and a few other things):

https://github.com/explosion/spaCy/blob/800737b41632c4b8dc24c992c8dd36f073549579/.github/azure-steps.yml#L67-L106

**Reply by iliakur** [reply]
Ok, let me know. The refactoring that occurred to me would take advantage of the `run_command` utility function. We don't have to test the download itself because that's essentially delegated to pip. We could just test that we generate the right pip command.

So `download` becomes (with args simplified for the sketch):

```python
def download(*args, **kwargs):
    run_command(make_download_cmd(*args, **kwargs))
```
The main logic is then inside `make_download_cmd` which we can easily test because it only returns a sequence of strings, doesn't actually hit the network in any way.

The current behavior is definitely not intuitive and encourages opaque invocations of the function.

**Reply by iliakur** [reply]
I can confirm that passing `direct` and `sdist` as positional arguments does indeed work. Thanks for that :)

**Comment by iliakur** [comment]
@adrianeboyd just chiming in to let you know that modifying `sys.path` before calling `spacy.load` works fine!
